File: internal/wal/wal.go
// Package wal handles Write-Ahead Log operations for Tidepool.
package wal

import (
	"bufio"
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"path"
	"strings"
	"time"

	"github.com/google/uuid"

	"github.com/tidepool/tidepool/internal/document"
	"github.com/tidepool/tidepool/internal/storage"
)

// Writer handles writing documents to WAL files.
type Writer struct {
	storage   *storage.Client
	namespace string
}

// NewWriter creates a new WAL writer.
func NewWriter(storage *storage.Client, namespace string) *Writer {
	return &Writer{
		storage:   storage,
		namespace: namespace,
	}
}

// Entry represents a single WAL entry.
type Entry struct {
	Timestamp time.Time          `json:"timestamp"`
	Document  *document.Document `json:"document"`
}

// Write appends documents to a new WAL file.
func (w *Writer) Write(ctx context.Context, docs []document.Document) (string, error) {
	if len(docs) == 0 {
		return "", nil
	}

	date := time.Now().UTC().Format("2006-01-02")
	walID := uuid.New().String()
	walPath := storage.WALPath(w.namespace, date, walID)

	var buf bytes.Buffer
	encoder := json.NewEncoder(&buf)

	for _, doc := range docs {
		entry := Entry{
			Timestamp: time.Now().UTC(),
			Document:  &doc,
		}
		if err := encoder.Encode(entry); err != nil {
			return "", fmt.Errorf("failed to encode WAL entry: %w", err)
		}
	}

	if err := w.storage.Put(ctx, walPath, buf.Bytes()); err != nil {
		return "", fmt.Errorf("failed to write WAL file: %w", err)
	}

	return walPath, nil
}

// Reader handles reading WAL files.
type Reader struct {
	storage   *storage.Client
	namespace string
}

// NewReader creates a new WAL reader.
func NewReader(storage *storage.Client, namespace string) *Reader {
	return &Reader{
		storage:   storage,
		namespace: namespace,
	}
}

// ListWALFiles returns all WAL files for the namespace.
func (r *Reader) ListWALFiles(ctx context.Context) ([]string, error) {
	prefix := storage.WALPrefix(r.namespace)
	keys, err := r.storage.List(ctx, prefix)
	if err != nil {
		return nil, fmt.Errorf("failed to list WAL files: %w", err)
	}

	// Filter to only include .jsonl files
	var walFiles []string
	for _, key := range keys {
		if strings.HasSuffix(key, ".jsonl") {
			walFiles = append(walFiles, key)
		}
	}
	return walFiles, nil
}

// ReadWALFile reads all entries from a WAL file.
func (r *Reader) ReadWALFile(ctx context.Context, walPath string) ([]Entry, error) {
	data, err := r.storage.Get(ctx, walPath)
	if err != nil {
		return nil, fmt.Errorf("failed to read WAL file %s: %w", walPath, err)
	}

	var entries []Entry
	scanner := bufio.NewScanner(bytes.NewReader(data))
	for scanner.Scan() {
		var entry Entry
		if err := json.Unmarshal(scanner.Bytes(), &entry); err != nil {
			return nil, fmt.Errorf("failed to parse WAL entry: %w", err)
		}
		entries = append(entries, entry)
	}

	if err := scanner.Err(); err != nil {
		return nil, fmt.Errorf("error scanning WAL file: %w", err)
	}

	return entries, nil
}

// ReadAllWALFiles reads all WAL entries from all WAL files.
func (r *Reader) ReadAllWALFiles(ctx context.Context) ([]Entry, []string, error) {
	walFiles, err := r.ListWALFiles(ctx)
	if err != nil {
		return nil, nil, err
	}

	var allEntries []Entry
	for _, walFile := range walFiles {
		entries, err := r.ReadWALFile(ctx, walFile)
		if err != nil {
			return nil, nil, err
		}
		allEntries = append(allEntries, entries...)
	}

	return allEntries, walFiles, nil
}

// DeleteWALFile deletes a WAL file.
func (r *Reader) DeleteWALFile(ctx context.Context, walPath string) error {
	return r.storage.Delete(ctx, walPath)
}

// WALInfo represents information about a WAL file.
type WALInfo struct {
	Path      string
	Date      string
	EntryCount int
}

// GetWALInfo parses WAL file path to extract info.
func GetWALInfo(walPath string) WALInfo {
	parts := strings.Split(walPath, "/")
	var date string
	if len(parts) >= 2 {
		date = parts[len(parts)-2]
	}
	return WALInfo{
		Path: walPath,
		Date: date,
	}
}

// ExtractDate extracts the date from a WAL path.
func ExtractDate(walPath string) string {
	// Path format: namespaces/{namespace}/wal/{date}/{uuid}.jsonl
	dir := path.Dir(walPath)
	return path.Base(dir)
}

File: crates/tidepool-common/src/wal.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::document::Document;
use crate::storage::{wal_path, wal_prefix, Store, StorageError};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Entry {
    pub ts: DateTime<Utc>,
    pub op: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub doc: Option<Document>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub delete_ids: Vec<String>,
}

#[derive(Clone)]
pub struct Writer<S: Store> {
    storage: S,
    namespace: String,
}

impl<S: Store> Writer<S> {
    pub fn new(storage: S, namespace: impl Into<String>) -> Self {
        Self {
            storage,
            namespace: namespace.into(),
        }
    }

    pub async fn write_upsert(&self, docs: Vec<Document>) -> Result<Option<String>, StorageError> {
        if docs.is_empty() {
            return Ok(None);
        }

        let date = Utc::now().format("%Y-%m-%d").to_string();
        let wal_id = Uuid::new_v4().to_string();
        let wal_path = wal_path(&self.namespace, &date, &wal_id);

        let mut buf = Vec::new();
        for doc in docs {
            let entry = Entry {
                ts: Utc::now(),
                op: "upsert".to_string(),
                doc: Some(doc),
                delete_ids: Vec::new(),
            };
            let mut line = serde_json::to_vec(&entry)
                .map_err(|err| StorageError::Other(format!("encode WAL entry: {}", err)))?;
            line.push(b'\n');
            buf.extend_from_slice(&line);
        }

        self.storage.put(&wal_path, buf).await?;
        Ok(Some(wal_path))
    }

    pub async fn write_delete(&self, ids: Vec<String>) -> Result<Option<String>, StorageError> {
        if ids.is_empty() {
            return Ok(None);
        }

        let date = Utc::now().format("%Y-%m-%d").to_string();
        let wal_id = Uuid::new_v4().to_string();
        let wal_path = wal_path(&self.namespace, &date, &wal_id);

        let entry = Entry {
            ts: Utc::now(),
            op: "delete".to_string(),
            doc: None,
            delete_ids: ids,
        };
        let mut buf = serde_json::to_vec(&entry)
            .map_err(|err| StorageError::Other(format!("encode WAL entry: {}", err)))?;
        buf.push(b'\n');

        self.storage.put(&wal_path, buf).await?;
        Ok(Some(wal_path))
    }
}

#[derive(Clone)]
pub struct Reader<S: Store> {
    storage: S,
    namespace: String,
}

impl<S: Store> Reader<S> {
    pub fn new(storage: S, namespace: impl Into<String>) -> Self {
        Self {
            storage,
            namespace: namespace.into(),
        }
    }

    pub async fn list_wal_files(&self) -> Result<Vec<String>, StorageError> {
        let prefix = wal_prefix(&self.namespace);
        let mut keys = self.storage.list(&prefix).await?;
        keys.retain(|k| k.ends_with(".jsonl"));
        keys.sort();
        Ok(keys)
    }

    pub async fn read_wal_file(&self, wal_path: &str) -> Result<Vec<Entry>, StorageError> {
        let data = self.storage.get(wal_path).await?;
        let mut entries = Vec::new();
        for line in data.split(|b| *b == b'\n') {
            if line.is_empty() {
                continue;
            }
            let entry: Entry = serde_json::from_slice(line)
                .map_err(|err| StorageError::Other(format!("parse WAL entry: {}", err)))?;
            entries.push(entry);
        }
        Ok(entries)
    }

    pub async fn read_all_wal_files(&self) -> Result<(Vec<Entry>, Vec<String>), StorageError> {
        let wal_files = self.list_wal_files().await?;
        let mut all_entries = Vec::new();
        for wal_file in &wal_files {
            let entries = self.read_wal_file(wal_file).await?;
            all_entries.extend(entries);
        }
        Ok((all_entries, wal_files))
    }

    pub async fn delete_wal_file(&self, wal_path: &str) -> Result<(), StorageError> {
        self.storage.delete(wal_path).await
    }
}

pub fn extract_date(wal_path: &str) -> Option<String> {
    let parts: Vec<&str> = wal_path.split('/').collect();
    if parts.len() < 5 {
        return None;
    }
    let date = parts.get(parts.len().saturating_sub(2))?;
    Some((*date).to_string())
}
